src/lsm/grid.zig

```diff
@@ -432,7 +425,27 @@ pub fn GridType(comptime Storage: type) type {
                 .completion = undefined,
                 .block = block,
             };
+
+            // Collect the current Read and any other matching/pending Reads to this IOP.
+            // There could be reads in the read_queue for the same address that our read
+            // finally got an IOP for. If we don't gather them here, they will eventually be 
+            // processed at the end of read_block_callback() but that will issue a new call to
+            // read_sectors().
             iop.reads.push(read);
+            {
+                // Make a copy here to avoid an infinite loop from pending_reads being 
+                // re-added to read_queue after not matching the current read. 
+                var copy = grid.read_queue;
+                grid.read_queue = .{};
+                while (copy.pop()) |pending_read| {
+                    if (pending_read.address == read.address) {
+                        assert(pending_read.checksum == read.checksum);
+                        iop.reads.push(pending_read);
+                    } else {
+                        grid.read_queue.push(pending_read);
+                    }
+                } 
```

Comment:
Instead of resetting the list and rebuilding it from scratch, could we instead iterate it and use `read_queue.remove(...)` to remove reads for the same address?

Reply:
It was my first though, but FIFO's are singly-linked list instead of doubly-linked lists; Removing nodes in the middle is O(n). I then thought I could special case the code in FIFO to check for a `prev` field and allow O(1) remove, but it meant modifying the inserts as well to keep `prev/next` updated (which didn't seem worth the extra code review). `remove()` would be ideal, but I figured this isn't too bad since it's already traversing.

Reply:
We are already iterating the full list anyway — couldn't you implement remove inline here, just by holding onto the prior element between iterations?

Reply:
(Assuming that ends up being simpler anyway, but I think it might).

Reply:
Id imagine it would look something like this
```zig
var prev: ?*Read = null;
var it: ?*Read = grid.read_queue.peek();

while (it) |pending_read| {
    it = pending_read.next;

    if (pending_read.address != read.address) {
        prev = pending_read;
        continue;
    }
    
    if (prev) |p| {
        var link = &p.next; // middle of FIFO
        if (it == null) list = &grid.read_queue.out; // last in FIFO
        link.* = it;
    } else { 
        _ = grid.read_queue.pop(); // first in FIFO
    }

    assert(pending_read.checksum == read.checksum);
    pending_read.next = null;
    iops.reads.push(pending_read);
}
```

Reply:
Alright, thanks for checking.